File: examples/lmm_pastes.R
library(lme4)

# 1. Load the dataset
data <- read.csv("tests/data/pastes.csv")
data$batch <- as.factor(data$batch)
data$cask <- as.factor(data$cask)

# 2. Fit the model
cat("\nFitting model: strength ~ 1 + (1 | batch/cask)\n")
fit <- lmer(strength ~ 1 + (1 | batch/cask), data = data, REML = TRUE)

# 3. Print the summary
cat("\n=== Model Summary ===\n")
print(summary(fit))

# 4. Generate Predictions
cat("\n=== Predictions ===\n")
cat("Generating predictions for population-level (re.form = NA)...\n")

newdata <- data.frame(
  batch = factor(c("A", "B", "C"), levels = levels(data$batch)),
  cask = factor(c("a", "b", "c"), levels = levels(data$cask))
)

preds <- predict(fit, newdata, re.form = NA)
cat("Predictions:\n")
print(preds)

File: examples/lmm_pastes.rs
use polars::prelude::*;
use lme_rs::lmer;

fn main() -> anyhow::Result<()> {
    // 1. Load the dataset
    let mut file = std::fs::File::open("tests/data/pastes.csv")?;
    let df = CsvReadOptions::default()
        .with_has_header(true)
        .into_reader_with_file_handle(&mut file)
        .finish()?;

    // Cast string columns
    let df = df
        .lazy()
        .with_columns(vec![
            col("batch").cast(DataType::String),
            col("cask").cast(DataType::String),
        ])
        .collect()?;

    println!("Fitting model: strength ~ 1 + (1 | batch/cask)");

    // 2. Fit the model
    let formula = "strength ~ 1 + (1 | batch/cask)";
    let fit = lmer(formula, &df, true)?;

    // 3. Print the summary
    println!("\n=== Model Summary ===");
    println!("{}", fit);

    println!("\n=== Predictions ===");
    println!("Generating predictions for population-level...");

    // 4. Generate Predictions
    let new_batch = Series::new("batch".into(), &["A", "B", "C"]);
    let new_cask = Series::new("cask".into(), &["a", "b", "c"]);

    let newdata = DataFrame::new(vec![new_batch.into(), new_cask.into()])?;

    let preds = fit.predict(&newdata)?;
    println!("Predictions:\n{:?}", preds);

    Ok(())
}
